### PR TITLE
feat(manager/asdf): add additional asdf supported tools

### PR DESCRIPTION
**Pull request by maennchen**
## Changes

Adds support for the following tools to the `asdf` manager:
awscli, bun, cargo-make, clojure, crystal, deno, direnv, dprint, elixir, elm, erlang, gauche, golang, haskell, helmfile, helm, hugo, idris, java, julia, just, kotlin, kustomize, lua, nim, nodejs, ocaml, perl, php, python, ruby, rust, scala, shellcheck, shfmt, terraform, trivy, zig

## Context

Currently, only `nodejs` is supported.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Test Repo: https://github.com/jshmrtn/renovate-asdf-test

Code is largely extracted from:
https://github.com/kachick/renovate-config-asdf/tree/1.8.0/plugins

Thanks @kachick!

### CODE REVIEW

File: lib/modules/manager/asdf/readme.md
Comment (rarkins) [inline lib/modules/manager/asdf/readme.md:6]:
```suggestion
The following tools are currently supported:
```

**Review by rarkins — CHANGES_REQUESTED** [review]
LGTM

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (viceice) [inline lib/modules/manager/asdf/upgradeable-tooling.ts:136]: should use our adopt java datasource instead

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (viceice) [inline lib/modules/manager/asdf/upgradeable-tooling.ts:227]: why does v3 uses a different datasource?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (maennchen) [reply lib/modules/manager/asdf/upgradeable-tooling.ts:227]: @kachick What is the reason for the two sources in the asdf config preset?

Comment (kachick) [reply lib/modules/manager/asdf/upgradeable-tooling.ts:227]: As far as I know, they have different repository for scala 2 and scala 3.
Official doc has the reference as https://github.com/scala/docs.scala-lang/blob/da2e086f5d684f850243af01fe98097a0ebc605f/scala3/reference/README.md?plain=1#L1-L6

However I'm not a scala user, it might be an misunderstanding 🤔

Comment (maennchen) [reply lib/modules/manager/asdf/upgradeable-tooling.ts:239]: The `scala/scala` repo does not have any major version 3 tags: https://github.com/scala/scala/tags

So I believe that is correct.

File: lib/modules/manager/asdf/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/asdf/extract.spec.ts:235]:
```suggestion
            versioning: 'semver',
```

see other comment

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (multani) [inline lib/modules/manager/asdf/upgradeable-tooling.ts:262]:
```suggestion
    datasource: GithubReleasesDatasource.id,
```

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (multani) [inline lib/modules/manager/asdf/upgradeable-tooling.ts:256]:
```suggestion
    datasource: GithubReleasesDatasource.id,
```

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (multani) [inline lib/modules/manager/asdf/upgradeable-tooling.ts:168]:
```suggestion
    datasource: GithubReleasesDatasource.id,
```

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (multani) [inline lib/modules/manager/asdf/upgradeable-tooling.ts:101]:
```suggestion
    datasource: GithubReleasesDatasource.id,
```

**Review by multani — CHANGES_REQUESTED** [review]
At least these tools are fetched from GitHub releases: using the GitHub tag can create a race condition where the tag has been published but not the release yet, which will fail asdf it the update is merged.

I implemented these on @kachick repository and that was an oversight from me, if we can have the fixed version here, that would be great!

**Comment by viceice** [comment]
> At least these tools are fetched from GitHub releases: using the GitHub tag can create a race condition where the tag has been published but not the release yet, which will fail asdf it the update is merged.
> 
> I implemented these on @kachick repository and that was an oversight from me, if we can have the fixed version here, that would be great!

sounds good. we should use releases when they are used by those tool instead of git tags

**Comment by multani** [comment]
> > At least these tools are fetched from GitHub releases: using the GitHub tag can create a race condition where the tag has been published but not the release yet, which will fail asdf it the update is merged.
> > I implemented these on @kachick repository and that was an oversight from me, if we can have the fixed version here, that would be great!
> 
> sounds good. we should use releases when they are used by those tool instead of git tags

Yes, I guess it could be applied to other tools, but it's probably plugin dependent, so in theory, it should be configured on a case by case basis ...

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (multani) [inline lib/modules/manager/asdf/upgradeable-tooling.ts:44]: https://github.com/asdf-community/asdf-crystal/blob/f8347c072a53a55a89d25f59e2ab9e8990a566f7/bin/install#L39
```suggestion
    datasource: GithubReleasesDatasource.id,
```

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (multani) [inline lib/modules/manager/asdf/upgradeable-tooling.ts:83]: Should be a GitHub Release, no? https://github.com/sakuro/asdf-gauche/blob/71f574d7c934bc4977eec4fb8005e2ecd0d7ed0c/bin/install#L32

File: lib/modules/manager/asdf/readme.md
Comment (multani) [inline lib/modules/manager/asdf/readme.md:23]:
```suggestion
- [helm](https://github.com/Antiarchitect/asdf-helm)
- [helmfile](https://github.com/feniix/asdf-helmfile)
```

File: lib/modules/manager/asdf/readme.md
Comment (multani) [inline lib/modules/manager/asdf/readme.md:15]: The URL doesn't work
```suggestion
- [dprint](https://github.com/asdf-community/asdf-dprint)
```

**Review by multani — DISMISSED** [review]
I had a look at all the links posted and make further adjustments.

Otherwise, this looks great!

Worst case scenario:

* The config references a git tag instead of a release: there may be a slight race condition and the new version can't be downloaded, but I guess it really happens for a short time.
* The config references a GitHub Release but only git tags are published: no updates will be proposed and Renovate's own config will have to be fixed. No harms done, bugs happen :)

Also, this is really tied up to the current implementation of each plugin and how upstream packages are released, so :shrug: 

Thanks a lot for importing all of this!

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (maennchen) [reply lib/modules/manager/asdf/upgradeable-tooling.ts:83]: Not sure: https://github.com/kachick/renovate-config-asdf/blob/main/plugins/gauche.json5#L11

Comment (multani) [reply lib/modules/manager/asdf/upgradeable-tooling.ts:83]: Hum, alright :sweat_smile: :+1:

File: lib/modules/manager/asdf/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/asdf/readme.md:8]: I don't want to maintain this long list myself. Can we automate "updating the list"? The code already knows what's supported somehow.

Comment (maennchen) [reply lib/modules/manager/asdf/readme.md:8]: @HonkingGoose If it is ok for you, I would offer to add a README generator as a new separate PR after this one is merged. Is that ok for you?

Comment (HonkingGoose) [reply lib/modules/manager/asdf/readme.md:8]: I'll let the maintainers decide what to do. 😉

Comment (maennchen) [reply lib/modules/manager/asdf/readme.md:8]: I've prepared a follow-up PR:

https://github.com/jshmrtn/renovate/compare/extend-asdf-support...jshmrtn:renovate:asdf_readme_generator

To avoid complicating matters, I will open that one as soon as this one is merged if no maintainer disagrees with this decision.

Comment (viceice) [reply lib/modules/manager/asdf/readme.md:8]: all tooling needs to be inside `tools` folder, so it doesn't get compiled to the npm package.

Comment (maennchen) [reply lib/modules/manager/asdf/readme.md:8]: @viceice Is this something that you want to adress in this PR already or is it ok for an upcoming PR?

Is there something else that needs to be resolved before merging this PR?

Comment (viceice) [reply lib/modules/manager/asdf/readme.md:8]: yes, seperate PR is ok

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/asdf/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/asdf/readme.md:6]: You'll have to fix the merge conflicts in this file. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by multani — DISMISSED** [review]

File: lib/modules/manager/asdf/extract.ts
Comment (viceice) [inline lib/modules/manager/asdf/extract.ts:20]:
```suggestion
    const depName = groups.toolName.trim();
```

File: lib/modules/manager/asdf/extract.ts
Comment (viceice) [inline lib/modules/manager/asdf/extract.ts:23]:
```suggestion
    const toolConfig = upgradeableTooling[depName];
```

File: lib/modules/manager/asdf/extract.ts
Comment (viceice) [inline lib/modules/manager/asdf/extract.ts:30]:
```suggestion
        depName,
```

File: lib/modules/manager/asdf/extract.ts
Comment (viceice) [inline lib/modules/manager/asdf/extract.ts:40]:
```suggestion
        depName,
```

**Review by viceice — COMMENTED** [review]
some small improvements.
needs coverage fix, you should add new small tests for thos cases

File: lib/modules/manager/asdf/readme.md
Comment (maennchen) [reply lib/modules/manager/asdf/readme.md:4]: merged :+1:

**Comment by maennchen** [comment]
@viceice I added the tests :+1:

File: lib/modules/manager/asdf/upgradeable-tooling.ts
Comment (viceice) [inline lib/modules/manager/asdf/upgradeable-tooling.ts:5]: please ensure they match the supported datasources export inside index file

Comment (maennchen) [reply lib/modules/manager/asdf/upgradeable-tooling.ts:5]: @viceice Done. I also added a test for it.

Comment (viceice) [reply lib/modules/manager/asdf/upgradeable-tooling.ts:5]: please read the contribution guide

https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#resolve-review-comments-instead-of-commenting

File: lib/modules/manager/asdf/index.spec.ts
Comment (maennchen) [reply lib/modules/manager/asdf/index.spec.ts:19]: @viceice Yes and no:

`PackageDependency<T>` defines the `datasource` as optional:

https://github.com/renovatebot/renovate/blob/9b5dd8c4f768747cdf83772925905f54082afe5f/lib/modules/manager/types.ts#L145

We're reusing that type here:

https://github.com/jshmrtn/renovate/blob/4d1bed2987696bc1fadbee257867ae72a2b009c4/lib/modules/manager/asdf/upgradeable-tooling.ts#L12

Therefore typescript considers it optional even though it is always set in practice.

Comment (viceice) [reply lib/modules/manager/asdf/index.spec.ts:19]: make `datasource`, `versioning` and `packageName` mandatory inside `StaticTooling`

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
next time you should open the PR from personal repo, otherwise we can't update your branch, with is needed for merge

**Comment by maennchen** [comment]
@viceice Would you prefer for me to open a new PR from my personal account to replace this one?

**Comment by viceice** [comment]
no, just use the github update button, and hope we get this merged fast enough before next main branch commit

**Pull request merged**
